rtl/eclrtl/rtlkey.cpp

```diff
@@ -815,56 +826,88 @@ class CIndirectKeySegmentMonitor : implements IKeySegmentMonitor, public CInterf
     unsigned offset;
 };
 
-//The base monitor provided to this segment monitor is constructed with offsets of 0
-class CVarOffsetKeySegmentMonitor : public CIndirectKeySegmentMonitor
+// The base monitor provided to this segment monitor is constructed with offsets of 0
+// offset refers to where the field would be in an "expanded" version of the record (all variable size - and thus unkeyed - fields being assumed null)
+
+class CNewVarOffsetKeySegmentMonitor : public CIndirectKeySegmentMonitor
 {
 public:
-    CVarOffsetKeySegmentMonitor(IKeySegmentMonitor * _base, unsigned _offset, IKeySegmentOffsetTranslator * _offsetTranslator) : CIndirectKeySegmentMonitor(_base, _offset) { offsetTranslator.setown(_offsetTranslator); }
+    CNewVarOffsetKeySegmentMonitor(IKeySegmentMonitor * _base, unsigned _offset, unsigned _fieldIdx)
+    : CIndirectKeySegmentMonitor(_base, _offset), fieldIdx(_fieldIdx)
+    {
+    }
 
-    CVarOffsetKeySegmentMonitor(MemoryBuffer &mb)
-        : CIndirectKeySegmentMonitor(mb)
+    CNewVarOffsetKeySegmentMonitor(MemoryBuffer &mb)
+    : CIndirectKeySegmentMonitor(mb)
     {
-        throwUnexpected();
+        mb.read(fieldIdx);
+    }
+
+    virtual bool matches(const RtlRow *keyval) const
+    {
+        return base->matchesBuffer(getSegmentBase(keyval));
     }
 
-    virtual bool matches(const void *keyval) const
+    virtual unsigned queryFieldNo() const { return fieldIdx+1; }
```

Comment:
Shouldn't this be
    virtual unsigned numFieldsRequired() const
an override attribute would help.